Added a file auditor and audit replay service

jwomeara opened the pull request:
1) This PR adds the ability to save audit messages as ~~URL-encoded~~, JSON-formatted lines to either a local file, or a file in HDFS in the event that we are unable to successfully publish the message to RabbitMQ.  

2) This PR adds audit replay endpoints and functionality, which can be used to either read lines from a local file, or a file from HDFS, and then parse the ~~URL-encoded~~, json-formatted audit messages contained within that file, and replay them via the audit controller.

jwomeara commented:
@brianloss @ivakegg @FineAndDandy I think this is ready for you guys to pick apart.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+ */
+@RestController
+@RequestMapping(path = "/v1/replay", produces = MediaType.APPLICATION_JSON_VALUE)
+@ConditionalOnProperty(name = "audit.replay.enabled", havingValue = "true")
```
Is there a reason we don't want this enabled by default? (matchIfMissing = true)? Also, most things in spring boot that do a conditionalonproperty don't do havingProperty of true, since that would exclude "True" for example. The default is for the value to not equal "false", which works well for something that is enabled by default and the user override is to disable.

jwomeara replied:
This is mostly just a personal preference of mine to have non-critical features disabled unless they are explicitly enabled.  I hadn't thought about the 'true' vs. 'True' bit...  This is definitely not the only place I'm using ConditionalOnProperty this way.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/AuditController.java:
```diff
@@ -171,22 +184,33 @@ public String audit(@RequestParam MultiValueMap<String,String> parameters) {
+            success = true;
+            try {
+                if (log.isDebugEnabled())
+                    log.debug("[" + auditParameters.getAuditId() + "] Attempting to log audit to the filesystem");
```
Same goes for basically all of the log messages you've added in this pr.

jwomeara replied:
Sure, this will be addressed in the next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+            }
+        } else {
+            resp = "No audit replay found with id " + id;
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
Usually when a resource is not found, a 404 (SC_NOT_FOUND) would be returned. Is there a reason for the bad request response?

jwomeara replied:
This will be addressed in the next commit.  I revamped the status codes across the board.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+                    resp = "Started audit replay with id " + id;
+                } else {
+                    resp = "Cannot start audit replay with id " + id;
+                    response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
Bad request is documented "The server cannot or will not process the request due to an apparent client error (e.g., malformed request syntax, size too large, invalid request message framing, or deceptive request routing)." Using this as a response code doesn't seem to fit here.

jwomeara replied:
The next commit will have this changed to SC_INTERNAL_SERVER_ERROR.

brianloss replied:
I'm sure there's a suitable one in https://en.wikipedia.org/wiki/List_of_HTTP_status_codes

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+                }
+            } else {
+                resp = "Cannot start audit replay with state " + status.getState();
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
Same as above.

jwomeara replied:
If the current state of the replay is not 'CREATED', then start is an invalid request.  Is there a better status code which reflects that?

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+        Status status = status(id, replayProperties.isPublishEvents());
+        
+        if (status == null)
+            response.sendError(HttpServletResponse.SC_BAD_REQUEST, "No audit replay found with id " + id);
```
Same comment as previous about sending a bad request here instead of a not found.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+        
+        List<Status> statuses = statusCache.retrieveAll();
+        if (statuses != null)
+            statuses = statuses.stream().map(status -> idleCheck(status, replayProperties.isPublishEvents())).collect(Collectors.toList());
```
It looks like idleCheck can never actually return a different status--it only has side effects. If that's the case, you could avoid duplicate list creation (and have more concise code) with:
```
if (statuses != null)
    statuses.forEach(status -> idleCheck(status, replayProperties.isPublishEvents());

jwomeara replied:
Good catch.  Updated in the next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+                update(status, sendRate, replayProperties.isPublishEvents());
+                resp = "Updated audit replay with id " + id;
+            } else {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
not found instead of bad request

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+        // update all of our replays.
+        List<Status> statuses = statusAll();
+        if (onlyUpdateRunning)
+            statuses = statuses.stream().filter(status -> status.getState() == RUNNING).collect(Collectors.toList());
```
I believe you an avoid the duplicate list creation with:
```
statuses.removeIf(status -> status.getState() != RUNNING)
```

jwomeara replied:
I didn't know about that.  Updated in the next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+            if (stop(status, replayProperties.isPublishEvents())) {
+                resp = "Stopped audit replay with id " + id;
+            } else {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
same comments as before about the response type

jwomeara replied:
If the audit replay is not running, and you try to stop it, is that not a bad request?  Should that be SC_FORBIDDEN or something like that instead?

brianloss replied:
Maybe a 409 fits better. All I know is the syntax of the request is fine, so 400 isn't the right answer.

jwomeara replied:
I am leaning towards 422 - UNPROCESSABLE_ENTITY as this implies that the server understands the request, it is syntactically correct, but was unable to process the instructions.  The error message I am sending along with this status code should make it clear what went wrong.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+    
+    private boolean stop(Status status, boolean publishEvent) {
+        // is the replay running?
+        if (status.getState() == RUNNING) {
```
This is a lot to parse, but I'm wondering if these status object can be updated by multiple threads? If so, then you need to synchronize here, yes?

jwomeara replied:
I thought about this a bit...  A status could be updated by multiple threads, or even by multiple audit services depending upon which pod you get when you hit the audit endpoint.  

To protect against race conditions I'll need some kind of distributed lock.  From the looks of it, Hazelcast has locking functionality built in.  I suppose I could either update the existing HazelcastCacheInspector, or extend it and add some methods which expose the locking functionality.  Thoughts?

jwomeara replied:
I'll add that given how I envision the replay endpoints being used, I doubt we would run encounter race condition issues with the code as-is, but I don't think it will be hard to utilize the Hazelcast locking functionality, and that would solve the problem.

jwomeara replied:
Ok, so the next commit will update the restcontroller to acquire a lock on the Hazelcast map before updating the status of a replay.  For running replays, updates and stop commands are forwarded to the running replay object, which will be responsible for updating the cache.

FineAndDandy replied:
this looks good

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+            if (resume(status)) {
+                resp = "Resumed audit replay with id " + id;
+            } else {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
response code

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+            }
+        } else {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            resp = "No audit replay found with id " + id;
```
not found

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+            }
+        } else {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            resp = "No audit replay found with id " + id;
```
not found

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+                statusCache.delete(status.getId());
+                resp = "Deleted audit replay with id " + id;
+            } else {
+                response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
response code

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+     */
+    @ApiOperation(value = "Deletes an audit replay.")
+    @RolesAllowed({"Administrator", "JBossAdministrator"})
+    @RequestMapping(path = "/{id}/delete", method = org.springframework.web.bind.annotation.RequestMethod.POST)
```
Usually, delete methods would use the DELETE verb. Is there a reason you aren't?

jwomeara replied:
Updated in the next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+     */
+    @ApiOperation(value = "Resumes an audit replay.")
+    @RolesAllowed({"Administrator", "JBossAdministrator"})
+    @RequestMapping(path = "/{id}/resume", method = org.springframework.web.bind.annotation.RequestMethod.POST)
```
The "PUT" verb is usually used for update/modification of a resource.

jwomeara replied:
Updated in the next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+     */
+    @ApiOperation(value = "Stops an audit replay.")
+    @RolesAllowed({"Administrator", "JBossAdministrator"})
+    @RequestMapping(path = "/{id}/stop", method = org.springframework.web.bind.annotation.RequestMethod.POST)
```
The "PUT" verb is usually used for update/modification of a resource.

jwomeara replied:
Updated in the next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+     */
+    @ApiOperation(value = "Updates all audit replays.")
+    @RolesAllowed({"Administrator", "JBossAdministrator"})
+    @RequestMapping(path = "/updateAll", method = org.springframework.web.bind.annotation.RequestMethod.POST)
```
The "PUT" verb is usually used for update/modification of a resource.

jwomeara replied:
Updated in the next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+     */
+    @ApiOperation(value = "Updates an audit replay.")
+    @RolesAllowed({"Administrator", "JBossAdministrator"})
+    @RequestMapping(path = "/{id}/update", method = org.springframework.web.bind.annotation.RequestMethod.POST)
```
The "PUT" verb is usually used for update/modification of a resource.

jwomeara replied:
Updated in the next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/ReplayController.java:
```diff
@@ -0,0 +1,726 @@
+     */
+    @ApiOperation(value = "Deletes all audit replays.")
+    @RolesAllowed({"Administrator", "JBossAdministrator"})
+    @RequestMapping(path = "/deleteAll", method = org.springframework.web.bind.annotation.RequestMethod.POST)
```
The "DELETE" verb would normally be used for deleting resources.

jwomeara replied:
Updated in the next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/config/ReplayConfig.java:
```diff
@@ -0,0 +1,39 @@
+@Configuration
+@EnableCaching
+@EnableConfigurationProperties(ReplayProperties.class)
+@ConditionalOnProperty(name = "audit.replay.enabled", havingValue = "true")
```
Same comments as before about the conditionalonproperty.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/config/ReplayProperties.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.concurrent.TimeUnit;
+
+@ConfigurationProperties(prefix = "audit.replay")
+public class ReplayProperties {
```
Maybe add validation (using validation annotations) to this? The millis values, I would guess at least all need to be positive, for example.

jwomeara replied:
added in next commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/config/ReplayConfig.java:
```diff
@@ -0,0 +1,39 @@
+public class ReplayConfig {
+    
+    @Bean
+    public ThreadPoolTaskExecutor auditReplayExecutor(ReplayProperties replayProperties) {
```
If the context is refreshed, this may cause thread leaks when this bean is re-created. May need to look into that.

jwomeara replied:
I'm working on a test for the next commit.

jwomeara replied:
I tested this.  As it is now, nothing happens on a refresh event.  However, if I mark the ThreadPoolTaskExecutor as @RefreshScope, any running replays are killed.  That's ok though, because eventually they will be marked as IDLE, at which point they can be resumed.  

For now, I am going to leave the RefreshScope annotation out.  What I will do instead is create a separate ticket to go over the entire audit service and try to get it configured to handle refresh appropriately.

ivakegg commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java:
```diff
@@ -0,0 +1,220 @@
+        }
+        
+        if (builder.fileUri != null)
+            fileSystem = FileSystem.get(new URI(builder.fileUri), config);
```
Lets use the basePath here instead.  Also suggest configuring basePath with the "fileUri" all in one property (e.g. file:///some/dir or hdfs://warehouse-a/some/dir)

jwomeara replied:
Working on this for the next commit.

ivakegg commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java:
```diff
@@ -0,0 +1,220 @@
+        this.maxFileAgeMillis = builder.maxFileAgeMillis;
+        
+        Configuration config = new Configuration();
+        config.set("fs.hdfs.impl", org.apache.hadoop.hdfs.DistributedFileSystem.class.getName());
```
This seems odd.  If we have the hadoop configuration mounted to the proper place, then "new Configuration()" will already load everything we need.

jwomeara replied:
Working on this for the next commit.

ivakegg commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java:
```diff
@@ -0,0 +1,220 @@
+        if (builder.fileUri != null)
+            fileSystem = FileSystem.get(new URI(builder.fileUri), config);
+        else
+            fileSystem = FileSystem.get(config);
```
this should never be allowed.  We always need a basePath (with protocol specified).  We need to fail if it is not specified.

ivakegg commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java:
```diff
@@ -0,0 +1,220 @@
+    protected long maxFileLenBytes;
+    protected long maxFileAgeMillis;
+    
+    protected FileSystem fileSystem;
```
Note that this is a Closeable object.  If this will exist for the lifetime of the jvm, then no problem.  Otherwise we need a close/cleanup mechanism.  A configuration refresh may be a path of concern here.

jwomeara replied:
It will live the lifetime of the container, so I'll leave it alone for now.

ivakegg commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java:
```diff
@@ -0,0 +1,220 @@
+            }
+            
+            currentDate = new Date();
+            currentFile = new Path(basePath, sdf.format(currentDate));
```
Can we use podname or hostname instead to avoid the possibility of conflicts?  The only reason I say this is hdfs can be troublesome as a locking mechanism.

jwomeara replied:
fixed in next commit

jwomeara commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java:
```diff
@@ -0,0 +1,220 @@
+        if (!fileSystem.exists(basePath))
+            fileSystem.mkdirs(basePath);
+        
+        this.sdf = new SimpleDateFormat("'" + builder.prefix + "'-yyyyMMdd_HHmmss.SSS'.json'");
```
The sdf created here is only used in the createNewFile method.  In the current implementation, the createNewFile method is only ever called if the calling thread is able to acquire a lock on the writeLock object.  I think this should be enough to prevent multiple threads from accessing the sdf.

brianloss commented on services/audit-service/pom.xml:
```diff
@@ -22,33 +24,91 @@
+                    </exclusion>
+                </exclusions>
+            </dependency>
+            <dependency>
```
spring-boot-starter-cache and datawave-in-memory-accumulo (test scope) are now included via spring-boot-starter-datawave so they can be removed from here.

jwomeara replied:
fixed in next commit

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/util/ConcurrentMapCacheInspector.java:
```diff
@@ -0,0 +1,61 @@
+package datawave.microservice.audit.replay.util;
```
hazelcast-common now includes a ConcurrentMapCacheInspector, so you might be able to drop this class.

jwomeara commented:
I rebased the branch against master and added a commit which should resolve all of the feedback that was given.  Let me know what you guys think when you get a chance.  @ivakegg @brianloss

FineAndDandy commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java:
```diff
@@ -0,0 +1,212 @@
+    
+    protected static final ObjectMapper mapper = new ObjectMapper();
+    
+    protected final SimpleDateFormat sdf;
```
since the create file call is wrapped inside the writeLock this is probably fine

jwomeara replied:
That was my thought as well.

FineAndDandy commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/config/FileAuditConfig.java:
```diff
@@ -0,0 +1,51 @@
+ *
+ */
+@Configuration
+@ConditionalOnProperty(name = "audit.auditors.file.enabled", havingValue = "true")
```
how is this different than the dump auditor?

jwomeara replied:
The file auditor our last ditch attempt at auditing, used when all other attempts to audit have failed (at the audit rest endpoint).  

The dump auditor is listening to the audit.dump queue, and writes any messages moved there to disk.  So, if you have some stubborn messages in your DLQ that you need to analyze, you can move them to the audit.dump queue, and then work with the file on disk.

FineAndDandy commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/config/ReplayProperties.java:
```diff
@@ -0,0 +1,146 @@
+    private long lockWaitTimeMillis = TimeUnit.SECONDS.toMillis(5);
+    
+    @PositiveOrZero
+    private long lockLeaseTimeMillis = TimeUnit.SECONDS.toMillis(5);
```
Can you better describe the difference between lease/lock timeouts? I'm not clear on when I would use one over the other and the impacts

jwomeara replied:
lockWaitTimeMillis is the amount of time we will wait to acquire a lock.

lockLeaseTimeMillis is the amount of time we are allowed to hold the lock before it is automatically released.  This is meant to safeguard against cases where an audit service instance goes down while processing a locking call.  If we didn't have this lease timeout, and our task died mid-call, the lock would never be released.

jwomeara replied:
Added a description in the latest commit.

brianloss commented on services/audit-service/pom.xml:
```diff
@@ -56,6 +125,13 @@
         </dependency>
     </dependencies>
     <repositories>
+        <repository>
```
I hate having 3rd party repos in here. Do we really need to depend on the Cloudera Hadoop version for what we're doing, or would it make more sense to depend on stock? Could see an argument both ways...

jwomeara replied:
I don't have a dog in this fight.  I was just trying to carry forward the precedent set in the warehouse pom.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/util/LockableCacheInspector.java:
```diff
@@ -0,0 +1,22 @@
+package datawave.microservice.audit.replay.util;
```
This doesn't seem audit-specific to me. Why isn't it in hazelcast-client?

jwomeara replied:
I have no objection to that.  Do you think it should still be a separate interface, or should I just add the methods to the CacheInspector interface?

brianloss replied:
I'd keep it separate. We don't want to require that every cache type supports locking functions.

jwomeara replied:
ok, cool.  that's how it is in the latest commit.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/util/LockableHazelcastCacheInspector.java:
```diff
@@ -0,0 +1,91 @@
+package datawave.microservice.audit.replay.util;
```
This doesn't seem audit-specific to me. Why isn't it in hazelcast-client?

jwomeara replied:
moved.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/replay/util/UniversalLockableCacheInspector.java:
```diff
@@ -0,0 +1,83 @@
+package datawave.microservice.audit.replay.util;
```
This doesn't seem audit-specific to me. Why isn't it in hazelcast-client?

jwomeara replied:
moved.

brianloss commented on services/sample_configuration/audit-dev.yml.example:
```diff
@@ -4,6 +4,11 @@
   port: 8743
 
+# Don't use a Hazelcast client, but rather use an embedded Hazelcast member. This is
+# only useful for development, and in production, this property shouldnot be set (or
```
"should not" instead of "shouldnot"

jwomeara replied:
updated in multiple places.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/AuditController.java:
```diff
@@ -171,22 +184,33 @@ public String audit(@RequestParam MultiValueMap<String,String> parameters) {
+        if (!success && fileAuditor != null) {
+            success = true;
+            try {
+                if (log.isDebugEnabled())
```
Unless the method call `auditParameters.getAuditId()` is expensive, there's no need to guard this statement with `if (log.isDebugEnabled())`

jwomeara replied:
updated.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java:
```diff
@@ -0,0 +1,212 @@
+        }
+    }
+    
+    protected String urlEncodeString(String value) {
```
Why the need for this? If the ObjectMapper won't do the URL encoding for you, then what's the purpose of catching the exception and returning the original value? If that case happens, then the object mapper will just fail on it. If the object mapper does take care of the encoding, then there's no need for this method.

jwomeara replied:
I'm just thinking ahead to the moment when I will be reading these json-formatted files in spark.  Spark natively supports reading json-formatted data, where each json object is written on a single line.

Multi-line json objects are also supported, but you have to write all of the json objects as an array, which isn't as convenient when writing in append mode (basically, it would be nice to avoid having to write beginning & closing square brackets).  

To get around this, I was planning on just URL-encoding the parameter values to ensure that no newlines would be present in a given json object.  This is easy to undo in spark when it comes time to process the data.  

To my knowledge, you don't just get URL encoding for free by using an object mapper, so that's why I have the extra step.  Also, while I do have logic for catching an unsupported encoding exception, you can see that it's hard coded to UTF8 (which is a supported character encoding), so I don't expect that exception to ever be thrown.  

Alternatively, I suppose I could write Base64 encoded bytes instead and URL encoding altogether, but i thought this would be good enough.  

What are your thoughts?

ivakegg replied:
The URL encoding sounds a like hoky to me.  In fact any nested encoding makes the file difficult to parse in other ways.  I would rather create an appropriate json output that can be readable by more than the spark processing.

ivakegg replied:
Sounds like some sort of escaping is required in any case, but using a standard JSON writer would avoid having to do anything like this.  I understand your Spark concerns, but I would rather have standardized json output that Spark can handle instead of having to do a urlDecode wherever we have to read these files.

jwomeara replied:
I had assumed that I would need to URL encode to ensure that the json stays as a single line (to play nicely with Spark), but after testing out the object mapper without doing any URL encoding, I think that will work just fine.  I updated the pr.

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java:
```diff
@@ -0,0 +1,212 @@
+    
+    protected void writeAudit(String jsonAuditParams) throws Exception {
+        OutputStream appendStream = (fileSystem instanceof LocalFileSystem) ? new FileOutputStream(new File(currentFile.toUri()), true)
+                        : fileSystem.append(currentFile);
```
Just curious... Why the complexity here on testing the file system instance? Does LocalFileSystem not support the append method?

jwomeara replied:
It does not, unfortunately.  :(

brianloss replied:
Hmm, weird. It implements the method. Oh well...

jwomeara replied:
During testing, it ultimately boiled down to this:
https://github.com/apache/hadoop/blob/release-2.6.0/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/ChecksumFileSystem.java#L352

brianloss commented on services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/config/FileAuditProperties.java:
```diff
@@ -0,0 +1,83 @@
+@Validated
+public class FileAuditProperties {
+    
+    @NotNull
```
`@NotBlank`, or is an empty string allowed for the pathUri?

jwomeara replied:
updated.
